nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/VerifyContentMAC.java

```diff
@@ -0,0 +1,187 @@
+        .name("SECRET_KEY")
+        .displayName("Secret key")
+        .description("Cryptographic key to calculate the hash")
+        .required(true)
+        .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+        .sensitive(true)
+        .build();
+
+    protected static final PropertyDescriptor MAC = new PropertyDescriptor.Builder()
+        .name("MAC")
+        .displayName("Message Authentication Code")
+        .description("The MAC to compare with the calculated value")
+        .required(true)
+        .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+        .expressionLanguageSupported(ExpressionLanguageScope.FLOWFILE_ATTRIBUTES)
+        .build();
+
+    protected static final Relationship FAILURE = new Relationship.Builder()
+        .name("failure")
+        .description("Signature Verification Failed")
+        .build();
+
+    protected static final Relationship SUCCESS = new Relationship.Builder()
+        .name("success")
+        .description("Signature Verification Succeeded")
+        .build();
+
+    private static final List<PropertyDescriptor> PROPERTIES = Collections.unmodifiableList(Arrays.asList(MAC_ALGORITHM, SECRET_KEY, MAC));
+    private static final Set<Relationship> RELATIONSHIPS = Collections.unmodifiableSet(new HashSet<>(Arrays.asList(SUCCESS, FAILURE)));;
+    private static final Charset CHARSET = StandardCharsets.UTF_8;
+    private static final int BUFFER_SIZE = 512000;
+    private static final String HEX_FORMAT = "%02x";
+    protected static final String MAC_CALCULATED_ATTRIBUTE = "mac.calculated";
+    protected static final String MAC_ALGORITHM_ATTRIBUTE = "mac.algorithm";
+
+    private SecretKeySpec secretKeySpec;
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return PROPERTIES;
+    }
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return RELATIONSHIPS;
+    }
+
+    @OnScheduled
+    public void setUp(ProcessContext context)  {
+        secretKeySpec = new SecretKeySpec(context.getProperty(SECRET_KEY).getValue().getBytes(CHARSET), HMAC_SHA256);
+    }
+
+    @Override
+    public void onTrigger(ProcessContext context, ProcessSession session) throws ProcessException {
+        FlowFile flowFile = session.get();
+
+        if (flowFile == null) {
+            return;
+        }
+
+        try {
+            String incomingMac = context.getProperty(MAC).evaluateAttributeExpressions(flowFile).getValue();
+            String algorithm = context.getProperty(MAC_ALGORITHM).getValue();
+            String calculatedMac = getCalculatedMac(session, flowFile, algorithm);
+
+            flowFile = enrichFlowFileAttributes(session, flowFile, algorithm, calculatedMac);
+
+            if (!incomingMac.equals(calculatedMac)) {
+                session.transfer(flowFile, FAILURE);
+            } else {
+                session.transfer(flowFile, SUCCESS);
+            }
```

Comment:
Recommend reversing this logic to check the positive case first, and use [MessageDigest.isEqual()](https://docs.oracle.com/javase/8/docs/api/java/security/MessageDigest.html#isEqual-byte:A-byte:A-) to compare byte arrays.

Reply:
done